File: Source/Matlab/teststream.m
% test subscribe

function teststream(host, modeflag)

  if(nargin < 1) host = 'localhost';
  end
  if(nargin < 2) modeflag = 1;
  end
  
% Sink
  snk = rbnb_sink(host,'mySink');
  cmapsnk = rbnb_cmap;
%  cmapsnk.Add('mySource/x');
  cmapsnk.Add('mySource/*');
  
  if(modeflag == 0) snk.Monitor(cmapsnk, 1);
  else              snk.Subscribe(cmapsnk)
  end
  
% Source
  maxwait = 10;
  ncache = 20;
  ngot = 0;
  nloop = ncache;
  src = rbnb_source(host, 'mySource', ncache, 'none', 0);
  
%   pause(.1)    % let Sink notice source?
  
  for i=1:nloop
    rbnb_put(src,'x',i, i, 1);
    pause(0.01)  % this should not be nec with sync'd put
%    if(i==1) pause(.5)  % grope
%    end

    for j=1:maxwait          % avoid forever loop
        cmapsnk = snk.Fetch(200);
        if(cmapsnk.NumberOfChannels > 0) break;
      end
      %fprintf('oops no data\n');
    end
    if(j >= maxwait) 
        warning('Gave up waiting to fetch data');
        continue;
    else
        ngot = ngot + 1;
    end
    
    gdat = cmapsnk.GetDataAsFloat64(0);
    gtim = cmapsnk.GetTimes(0);
    gnam = cmapsnk.GetName(0);
  
    if(strcmp(gnam, 'mySource/x') ~= 1) 
      error('Channel name mismatch');
    end
    if(size(gdat) < 1)
      error('Fetched wrong amount of data');
    end
    if(gdat ~= gtim)    % we put in matching data=time

      error('Channel data mismatch');
    end
    if(gdat ~= i)    % we put in matching data=i
      i
      error('Channel data mismatch');
    end
  end
  
% all done with RBNB access
  src.CloseRBNBConnection;
  snk.CloseRBNBConnection;
  
% fprintf('ngot: %g, nloop: %g, fraction: %g\n',ngot,nloop,ngot/nloop);
  if(ngot < nloop) 
    wtxt = sprintf('Lost %d/%d data frame(s)\n', nloop-ngot, nloop);
    if(ngot/nloop < .5) error(wtxt)
    else                fprintf(wtxt)
    end

  end

  if(modeflag == 0) fprintf('PASS: monitor stream test\n');
  else              fprintf('PASS: subscribe stream test\n');
  end
  
  pause(.2);    % make sure its closed (close should be synchronous)
return

File: Source/JUnitTesting/src/com/rbnb/mtest/teststream.java
// RBNB JUNIT Test Template
// test streaming data source with subscribe

package com.rbnb.mtest;

import static org.junit.Assert.assertEquals;

import org.junit.Test;

import com.rbnb.sapi.*;

public class teststream extends testrbnb {		
	String clientName = getClass().getSimpleName();

	@Test 
	public void thisTest() {
		assertEquals(clientName, true, runtest(clientName));
	}

	public boolean test(String host) throws SAPIException {
		int modeflag = 1;

		// Sink
		Sink snk = new Sink();
		snk.OpenRBNBConnection(host,"mySink");
		ChannelMap cmapsnk = new ChannelMap();
		cmapsnk.Add("mySource/*");

		if(modeflag == 0) snk.Monitor(cmapsnk, 1);
		else              snk.Subscribe(cmapsnk);

		// Source
		int maxwait = 10;
		int ncache = 20;
		int ngot = 0;
		int nloop = ncache;

		Source src = new Source(ncache,"none",0);
		src.OpenRBNBConnection(host, "mySource"); 

		ChannelMap cmap = new ChannelMap();
		int ix = cmap.Add("x");

		for(int i=0; i<nloop; i++) {
			cmap.PutTime((double)i, 1.);
			double[] d = { (double)i };
			cmap.PutDataAsFloat64(ix, d);
			src.Flush(cmap);
			
			int j=0;
			for (; j<maxwait; j++) {          // avoid forever loop
				cmapsnk = snk.Fetch(500);
//				System.err.println("cmapsnk: "+cmapsnk);
				if(cmapsnk.NumberOfChannels() > 0) break;
			}
			if(j >= maxwait) throw new SAPIException("Gave up waiting to fetch data"); 
			else  			 ngot = ngot + 1;

			double[] gdat = cmapsnk.GetDataAsFloat64(0);
			double[] gtim = cmapsnk.GetTimes(0);
			String gnam = cmapsnk.GetName(0);

			if(!gnam.equals("mySource/x")) throw new SAPIException("Channel name mismatch");
			if(gdat.length < 1) throw new SAPIException("Fetched wrong amount of data");

			if(gdat[0] != gtim[0]) throw new SAPIException("Channel data mismatch");   	// we put in matching data=time
			if(gdat[0] != i) throw new SAPIException("Channel data mismatch");   		// we put in matching data=i
		}

		// all done with RBNB access
		src.CloseRBNBConnection();
		snk.CloseRBNBConnection();

		// fprintf('ngot: %g, nloop: %g, fraction: %g\n',ngot,nloop,ngot/nloop);
		if(ngot < nloop) {
			String wtxt = "Lost "+(nloop-ngot)+"/"+nloop+" data frame(s)";
			if((double)ngot/(double)nloop < .5) throw new SAPIException(wtxt);
			else                				System.err.println(wtxt);
		}

		return true;			// fill in test logic here
	}
}
